workflows/vertex_classification/include/pando-wf1/gnn.hpp

```diff
@@ -0,0 +1,356 @@
+public:
+  constexpr GraphNeuralNetwork() noexcept = default;
+
+  void initialize(pando::GlobalPtr<Graph> dGraphPtr) {
+    this->numPXNs = pando::getPlaceDims().node.id;
+    PANDO_CHECK(this->nodeList_.initialize(this->numPXNs));
+    for (std::int16_t n = 0; n < this->numPXNs; ++n) {
+      this->nodeList_[n] =
+          galois::PlaceType{pando::Place{pando::NodeIndex{n}, pando::anyPod, pando::anyCore},
+                            pando::MemoryType::Main};
+    }
+    this->gnnGraphPtr_ = static_cast<decltype(this->gnnGraphPtr_)>(
+        pando::getDefaultMainMemoryResource()->allocate(sizeof(graph::GNNGraph<Graph>)));
+    this->gnnGraphPtr_->initialize(dGraphPtr);
+    this->InitializeLayers(dGraphPtr);
+    // Setup optimizers
+    pando::Vector<LayerDimension> optimizerSizes;
+    PANDO_CHECK(optimizerSizes.initialize(0));
+    PANDO_CHECK(optimizerSizes.pushBack(this->gnnGraphPtr_->VertexFeatureLength() * 16));
+    PANDO_CHECK(optimizerSizes.pushBack(16 * this->gnnGraphPtr_->GetNumClasses()));
+    this->optimizer_.initialize(optimizerSizes, this->numGCNLayers_);
+  }
+
+  ~GraphNeuralNetwork() {
+    for (std::uint32_t l = 0; l < this->numGCNLayers_; ++l) {
+      pando::getDefaultMainMemoryResource()->deallocate(
+          static_cast<pando::GlobalPtr<GraphConvolutionalLayer<Graph>>>(this->gcnLayers_[l]),
+          sizeof(GraphConvolutionalLayer<Graph>));
+    }
+  }
+
+  /**
+   * @brief Allocates and initialize layers.
+   */
+  void InitializeLayers(pando::GlobalPtr<Graph> dGraphPtr) {
+    galois::HostIndexedMap<pando::Array<GNNFloat>> backwardOutputMatrix;
+    struct Tpl {
+      LayerDimension inColumnLen;
+      LayerDimension outColumnLen;
+      pando::GlobalPtr<Graph> gPtr;
+    };
+
+    /***************************************************************************
+     *                        GCN initialization                           *
+     ***************************************************************************/
+
+    PANDO_CHECK(this->gcnLayers_.initialize(this->numGCNLayers_));
+    LayerDimension inColumnLen{0}, outColumnLen{0};
+    // Setup GCNlayers with AGILE's configurations
+    for (std::uint32_t l = 0; l < this->numGCNLayers_; ++l) {
+      // ** [Dimension] **
+      //
+      // ** 1st GCN Layer:
+      // Input dimension: (# of sampled vertices) x 30
+      // Output dimension: (# of sampled vertices) x 16
+      //
+      // ** 2nd GCN Layer:
+      // Input dimension: (# of sampled vertices) x 16
+      // Output dimension: (# of sampled vertices) x 5
+
+      if (l == 0) {
+        inColumnLen = this->gnnGraphPtr_->VertexFeatureLength();
+        outColumnLen = 16;
+      } else if (l == 1) {
+        inColumnLen = 16;
+        outColumnLen = this->gnnGraphPtr_->GetNumClasses();
+      }
+
+      // Per-host layer dimension; Each PXN materializes and uses different matrices
+      // with different dimensions
+      galois::HostIndexedMap<gnn::GNNLayerDimensions> dimension;
```

Comment:
Host Local Storage right?

Reply:
Same thing for this.

Reply:
What is difference between HostLocalStorage and HostIndexedMap? Let me check it but you have simple explanation?